modules/livestore/instance.go

```diff
@@ -0,0 +1,333 @@
+package livestore
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/google/uuid"
+	"github.com/grafana/tempo/modules/ingester"
+	"github.com/grafana/tempo/pkg/livetraces"
+	"github.com/grafana/tempo/pkg/model"
+	"github.com/grafana/tempo/pkg/tempopb"
+	v1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	"github.com/grafana/tempo/pkg/tracesizes"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/grafana/tempo/tempodb/encoding"
+	"github.com/grafana/tempo/tempodb/encoding/common"
+	"github.com/grafana/tempo/tempodb/wal"
+)
+
+type instance struct {
+	tenantID string
+	logger   log.Logger
+
+	// WAL and encoding
+	wal *wal.WAL
+	enc encoding.VersionedEncoding
+
+	// Block management
+	blocksMtx      sync.Mutex
+	headBlock      common.WALBlock
+	walBlocks      map[uuid.UUID]common.WALBlock
+	completeBlocks map[uuid.UUID]*ingester.LocalBlock
+	lastCutTime    time.Time
+
+	// Live traces
+	liveTracesMtx sync.Mutex
+	liveTraces    *livetraces.LiveTraces[*v1.ResourceSpans]
+	traceSizes    *tracesizes.Tracker
+
+	// Block offset metadata (set during coordinated cuts)
+	// blockOffsetMeta map[uuid.UUID]offsetMetadata // TODO: Used for checking data integrity
+
+	// Overrides
+	overrides Overrides
+}
+
+func newInstance(instanceID string, wal *wal.WAL, overrides Overrides, logger log.Logger) (*instance, error) {
+	enc := encoding.DefaultEncoding()
+
+	i := &instance{
+		tenantID:       instanceID,
+		logger:         log.With(logger, "tenant", instanceID),
+		wal:            wal,
+		enc:            enc,
+		walBlocks:      map[uuid.UUID]common.WALBlock{},
+		completeBlocks: map[uuid.UUID]*ingester.LocalBlock{},
+		liveTraces:     livetraces.New[*v1.ResourceSpans](func(rs *v1.ResourceSpans) uint64 { return uint64(rs.Size()) }, 30*time.Second, 5*time.Minute),
+		traceSizes:     tracesizes.New(),
+		overrides:      overrides,
+		// blockOffsetMeta:   make(map[uuid.UUID]offsetMetadata),
+	}
+
+	err := i.resetHeadBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	return i, nil
+}
+
+func (i *instance) pushBytes(ts time.Time, req *tempopb.PushBytesRequest) {
+	// For each pre-marshalled trace, we need to unmarshal it and push to live traces
+	for j, traceBytes := range req.Traces {
+		if j >= len(req.Ids) {
```

Comment:
Can this be moved outside this loop? `if len(req.Traces) != len(req.Ids)` or even higher in the chain?

Reply:
Yes.